### PR TITLE
Feat(Terraform): Baseline and Genai Scenarios

### PR DESCRIPTION
**Pull request by nidhirw**
## **What's included in this PR?**

This pull request introduces several significant changes to the project, primarily focusing on improving the development environment and adding comprehensive Terraform-based deployment guides and configurations for the 'Baseline' and 'GenAI' scenarios. Below is a summary of the most important changes:

### **Features**

**1. Terraform Templates for Baseline Scenario:**
 
- **Azure Resource Configuration for APIM**:
   - Introduced a new `networking.tf` file to manage Azure Virtual Networks, subnets, and Network Security Groups (NSGs).
   - Added service delegation for the deployment subnet.
   - Added new variables and outputs for easier reference to subnet and VNet IDs.
 
- **Monitoring Resources**:
   - Created `azmon.tf` for setting up `azurerm_log_analytics_workspace` and `azurerm_application_insights`.
   - Included lifecycle rules to prevent accidental deletion of critical monitoring resources.
   - Added outputs for `workspaceId`, `instrumentationKey`, and `keyVaultId`.
 
- **Private DNS Zone Module**:
   - Added a new module for `azurerm_private_dns_zone` and `azurerm_private_dns_zone_virtual_network_link`.
   - Provided flexibility with new variables for DNS zones and VNet linkage.
 
- **Private Endpoint Configuration**:
   - Created a module for `azurerm_private_endpoint` including private service connection and DNS zone group setup.
   - Introduced outputs for `id`, `private_dns_zone_group`, and `private_dns_zone_configs`.
 
- **Key Vault Setup**:
   - Added a Terraform configuration for creating and managing Key Vault, including access policies and private endpoints.
 
- **Provider and Backend Configuration**:
   - Introduced `provider.tf` for setting up Terraform providers and backend storage configurations.
 
**2. Terraform Templates for GenAI Scenario**

- **OpenAI Module**:
   - Added new modules for `azurerm_cognitive_account` and `azurerm_cognitive_deployment` for three openai instances and multiple openai model deployments. 
   - Provided flexibility with new variables for DNS zones and VNet linkage.

- **Private DNS Zone Module**:
   - Added a new module for `azurerm_private_dns_zone` and `azurerm_private_dns_zone_virtual_network_link`, that are utilised in openai private endpoints per openai instance.
   - Provided flexibility with new variables for DNS zones and VNet linkage.

- **Private Endpoint Configuration**:
   - Created a module for `azurerm_private_endpoint` including private service connection and DNS zone group setup, utilised for three openai instances.
   - Introduced outputs for `id`, `private_dns_zone_group`, and `private_dns_zone_configs`.

- **API OpenAI Policies**:
   - Created a module for `azurerm_api_management_product`, `azurerm_api_management_product_api` and three `azurerm_api_management_backend` deployments for each openai instance. 
   - Created `azurerm_api_management_policy_fragment` for each openai policy. 

- **EventHub**:
   - Created a module for `azurerm_eventhub` and `azurerm_eventhub_namespace` including a role assignment for API management. 

**3. Deployment Scripts:**
   - Refactored and relocated deployment scripts under `scenarios/scripts/bicep`.
   - Created deployment scripts for terraform templates, under `scenarios/scripts/terraform`. 
   - Updated paths in scripts to match the new directory structure.
 
### **Docs**
 
1. **Documentation Updates**:
   - Updated README files with correct paths and deployment instructions reflecting the new structure.
   - Added a Terraform-based deployment guide for both `apim-baseline` and `workload-genai` scenarios, including detailed steps and prerequisites.
 
### **Chore**
 
1. **Dev Container**
   - Added use of dev container.

2. **Terraform Sample Script for Terraform Backend**
   - Added the script `scenarios/scripts/terraform/azure-backend-sample.sh` to create the backend required for terraform.

3. **Code Formatting**:
   - Updated `main.bicep` and other related files to improve code readability and consistency.
   - Made minor formatting changes across scripts for better readability and maintainability.
 
## **How to review the PR**

**How to deploy the baseline scenario:**

- Follow the READ.me here: [scenarios/apim-baseline/terraform/README.md](https://github.com/Azure/apim-landing-zone-accelerator/pull/133/files#diff-aad19567ff6a788b10871db2566fbce7866309955c0571cf63f875533478f93d)

- After following the steps, you will have a `.env` and your terraform backend configuration in `${ENVIRONMENT_TAG}-backend.hcl` e.g. `dev-backend.hcl`.

- Run the following command to deploy the APIM baseline

    ```bash
    ./scripts/terraform/deploy-apim-baseline.sh -y
    ```

- At the end of the deployment, run the CURL command that is generated in the command line. 

**How to deploy the GenAI scenario:**

After running the baseline deployment script, run the following command to deploy the genai scenario:

    ```bash
    ./scripts/terraform/deploy-workload-genai.sh
    ```

- At the end of the deployment, run the CURL command that is generated in the command line.

### CODE REVIEW

**Review by seenu433 — APPROVED** [review]
The changes are good. Functions scenario is pending.

File: scenarios/apim-baseline/terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    | `CERT_TYPE` | selfsigned will create a self-signed certificate for the APPGATEWAY_FQDN. custom will use an existing certificate in pfx format that needs to be available in the [certs](../../certs) folder and named appgw.pfx | **selfsigned** | **custom** |
+    | `CERT_PWD` | The password for the pfx certificate. Only required if CERT_TYPE is custom. | **N/A** | **password123** |
+    | `RANDOM_IDENTIFIER` | Optional 3 character random string to ensure deployments are unique. Automatically assigned if not provided | **abc** | **pqr** |
+
+1. For terraform to work, you'll need to setup the [tf backend](https://developer.hashicorp.com/terraform/language/settings/backends/configuration). As part of the repository we provide a `azure-backend-sample.sh` script. This script will create a storage account and a container to store the terraform state. You can run the script with the following command:
+
+    ```bash
+    ./azure-backend-sample.sh \
+         --resource-group my-resource-group \
+         --storage-account my-storage-account \
```
Comment (prasann) [inline scenarios/apim-baseline/terraform/README.md:63]: please adjust the storage name without hyphens. helps to do a copy paste.

File: scenarios/apim-baseline/terraform/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
+# Azure API Management - Secure Baseline [Terraform]
+
+This is the Terraform-based deployment guide for [Scenario 1: Azure API Management - Secure Baseline](../README.md).
+
+## Prerequisites
+
```
Comment (prasann) [inline scenarios/apim-baseline/terraform/README.md:6]: I believe, one of the prerequisites is to have a terraform setup in the local machine.

File: scenarios/apim-baseline/terraform/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+   ```hcl
+   resource_group_name  = "my-resource-group"
+   storage_account_name = "mystorageaccount"
+   container_name       = "my-container"
+   ```
+
+
+1. Deploy the reference implementation.
+
+   Run the following command to deploy the APIM baseline
```
Comment (prasann) [inline scenarios/apim-baseline/terraform/README.md:78]: add a note to inform users that they will encounter prompts during the script execution and need to respond with `y` to proceed.

File: scenarios/workload-genai/terraform/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+## Steps
+
+Run the following command to deploy the scenarios
+
+```bash
+./scripts/terraform/deploy-workload-genai.sh
+```
+
+Test the hello api using hte generated command from the output
```
Comment (prasann) [inline scenarios/workload-genai/terraform/README.md:17]:
```suggestion
Test the hello api using the generated command from the output
```

**Review by prasann — CHANGES_REQUESTED** [review]
apart from these mentioned improvements, i faced couple of intermittent issues while running these scripts.

Issue1:
Post setting up the apim-baseline the script failed here. I see the token got expired, and the difference is about 7 mins. I'm not sure, whether i delayed somewhere or not. But in the next run, it went through.

```
Error: creating User Assigned Identity (Subscription: "xxx"
│ Resource Group Name: "rg-networking-lz00-dev-eastus-nuly"
│ Name: "identity-appgw-lz00-dev-eastus-nuly"): unexpected status 401 (401 Unauthorized) with error: ExpiredAuthenticationToken: The access token expiry UTC time '9/6/2024 9:07:31 AM' is earlier than current UTC time '9/6/2024 9:14:54 AM'.
│
│   with module.gateway.azurerm_user_assigned_identity.user_assigned_identity,
│   on modules/gateway/gateway.tf line 16, in resource "azurerm_user_assigned_identity" "user_assigned_identity":
│   16: resource "azurerm_user_assigned_identity" "user_assigned_identity" {
│
│ creating User Assigned Identity (Subscription: "xxx"
│ Resource Group Name: "rg-networking-lz00-dev-eastus-nuly"
│ Name: "identity-appgw-lz00-dev-eastus-nuly"): unexpected status 401 (401 Unauthorized) with error: ExpiredAuthenticationToken: The access token expiry UTC time '9/6/2024 9:07:31 AM' is earlier than current UTC time '9/6/2024 9:14:54 AM'.
```

Issue 2:

While running `deploy-workload-genai.sh` I got an error once, 

```
Error: creating Policy Fragment (Subscription: "xxx"
│ Resource Group Name: "rg-apim-tlz01-dev-australiaeast-boi"
│ Service Name: "apim-tlz01-dev-australiaeast-boi"
│ Policy Fragment Name: "simple-priority-weighted"): polling after CreateOrUpdate: executing request: unexpected status 404 (404 Not Found) with error: ResourceNotFound: PolicyFragment not found.
│
│   with module.apiManagement.azurerm_api_management_policy_fragment.simpleRoundRobinPolicyFragment,
│   on modules/apim_policies/apimanagement.tf line 88, in resource "azurerm_api_management_policy_fragment" "simpleRoundRobinPolicyFragment":
│   88: resource "azurerm_api_management_policy_fragment" "simpleRoundRobinPolicyFragment" {
```

I was able to locate this file in the mentioned path, and in the second run it did pass.

Just dropping a note on these 2 errors, in case if you have also faced these in the past try to figure a way out or add some workarounds to the troubleshooting guide.

Both the above mentioned issues didn't show up in the subsequent runs. So can't really pin point the issue.

**Review by aarthiem — APPROVED** [review]
Thanks @jsburckhardt for the Terraform updates!!

**Pull request merged**